pkg/resources/finalizers.go

```diff
@@ -0,0 +1,86 @@
+package resources
+
+import (
+	"context"
+
+	"github.com/integr8ly/integreatly-operator/pkg/apis/integreatly/v1alpha1"
+	oauthClient "github.com/openshift/client-go/oauth/clientset/versioned/typed/oauth/v1"
+	"github.com/sirupsen/logrus"
+	k8serr "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	pkgclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// FinalizeProducts is a map of the products to finalise when the installation is deleted
+var FinalizeProducts = make(map[string]*v1alpha1.InstallationProductStatus)
+
+// AddFinalizer adds a finalizer to the custom resource. This allows us to clean up oauth clients
+// and other cluster level objects owned by the installation before the cr is deleted
+func AddFinalizer(ctx context.Context, inst *v1alpha1.Installation, client pkgclient.Client, product *v1alpha1.InstallationProductStatus, finalizer string) error {
+	if !contains(inst.GetFinalizers(), finalizer) && inst.GetDeletionTimestamp() == nil {
+		inst.SetFinalizers(append(inst.GetFinalizers(), finalizer))
+		err := client.Update(ctx, inst)
+		if err != nil {
+			logrus.Error("Error adding finalizer to custom resource", err)
+			return err
+		}
+	}
+	if _, ok := FinalizeProducts[finalizer]; !ok {
+		FinalizeProducts[finalizer] = product
+	}
+	return nil
+}
+
+// RemoveOauthClient deletes an oauth client owned by a product and removes its finalizer from
+// the installation custom resource
+func RemoveOauthClient(ctx context.Context, inst *v1alpha1.Installation, client pkgclient.Client, oauthClient oauthClient.OauthV1Interface, finalizer string, oauthId string) error {
+	if contains(inst.GetFinalizers(), finalizer) {
```

Comment:
This finalizer code is repeated in the `ReconcileFinalizer` so it can be removed here.
This function should just delete the Oauth Client